### PR TITLE
Optimise Ip2Country

### PR DESCRIPTION
**Pull request by bemusementpark**

### CODE REVIEW

File: libsession/src/main/java/org/session/libsession/utilities/ListUtil.kt [hunk 1]
```diff
@@ -0,0 +1,28 @@
+package org.session.libsession.utilities
+
+/**
+ * @return the last element that satisfies the [predicate].
+ *
+ * This function has undefined behavior if:
```
Comment (ThomasSession) [inline libsession/src/main/java/org/session/libsession/utilities/ListUtil.kt:6]: Why 'undefined behaviour'?
Isn't the behaviour what you described above? Even if multiple are true, the last one will be returned, won't it?

Comment (ThomasSession) [reply libsession/src/main/java/org/session/libsession/utilities/ListUtil.kt:6]: Or do you mean:
```
 * This function has undefined behavior if:
 * - The given [predicate] will return [true] for some number of elements, and the elements are not sorted 
 *   according to the predicate. 
 ```

File: app/src/main/java/org/thoughtcrime/securesms/util/IP2Country.kt [hunk 2]
```diff
@@ -5,59 +5,47 @@ import android.content.Context
...
-    val countryNamesCache = mutableMapOf<String, String>()
+    val countryNamesCache = mutableMapOf<String, String?>()
 
-    private fun Ipv4Int(ip:String) = ip.takeWhile { it != '/' }.split('.').foldIndexed(0L) { i, acc, s ->
-        val asInt = s.toLong()
-        acc + (asInt shl (8 * (3-i)))
-    }
+    private fun ipv4Int(ip: String) =
+        ip.takeWhile { it != '/' }.split('.').reversed()
+            .fold(0L) { acc, octet -> acc shl 8 or octet.toLong() }
```
Comment (ThomasSession) [inline app/src/main/java/org/thoughtcrime/securesms/util/IP2Country.kt:29]: Those two functions are not equivalent.
You can test out their outputs and they will output different things.
Was the original broken? If not the new function is not the same.

File: app/src/main/java/org/thoughtcrime/securesms/util/IP2Country.kt [hunk 3]
```diff
@@ -66,64 +54,45 @@ class IP2Country private constructor(private val context: Context) {
...
-    private fun populateCacheIfNeeded() {
-        ThreadUtils.queue {
-            OnionRequestAPI.paths.iterator().forEach { path ->
-                path.iterator().forEach { snode ->
-                    cacheCountryForIP(snode.ip) // Preload if needed
-                }
+    /**
+     * get the country name for the given IP address.
+     *
+     * Cache the result even if null, as this will improve subsequent searches.
```
Comment (ThomasSession) [inline app/src/main/java/org/thoughtcrime/securesms/util/IP2Country.kt:86]: Are you sure this is true?
`getOrPut`: _Returns the value for the given key if the value is present **and not null**_
Won't null value always go through the put then?

**Comment by simophin** [comment]
We used to have fun optimising these ip country lookup. It turns out if you optimise the file storage to start with, with a fixed length binary data structure, like this
```
|--- 4 bytes IP start --- | --- 4 bytes IP end --- | --- 2 byte country code --- |
```

Then pre-sort the data by the IP start, so that 
1. You can do binary search on the list with the IP (IP is just 4 byte integer)
2. You can use `seek` to jump to any line in the file (because each line is fixed length, you'd be able to work it out easily), so you don't need to load all of them in the memory. (Or you can also use `mmap` to map the file into memory, instead of reading them all).

Doing this optimisation probably means there's no need to really cache any data, as it would be fast enough.

Just my two cents for some fun ideas though

**Comment by ThomasSession** [comment]
Closing this PR as it is still in draft.
Parts of this have been integrated into incremental changes we are doing to the ip to country feature.

**Pull request closed without merging**